src/pages/workspaces/workspace/notebooks/TerminalLauncher.js

```diff
@@ -46,33 +45,69 @@ const TerminalLauncher = _.flow(
     return Jupyter.notebooks(namespace, clusterName).setCookie()
   }
 
-  componentDidMount() {
-    this.loadIframe()
+  async startCluster() {
+    const { namespace, refreshClusters, ajax: { Jupyter } } = this.props
+    await refreshClusters()
+    if (!this.props.cluster) {
+      await Jupyter.cluster(namespace, Utils.generateClusterName()).create({
+        machineConfig: Utils.normalizeMachineConfig({})
+      })
+      this.setState({ createRequested: true })
+    }
+
+    while (this.mounted) {
+      await refreshClusters()
+      const cluster = this.props.cluster // Note: reading up-to-date prop
+      const status = cluster && cluster.status
+
+      if (status === 'Running') {
+        return
+      } else if (status === 'Stopped') {
+        const { googleProject, clusterName } = cluster
+        await Jupyter.cluster(googleProject, clusterName).start()
+        refreshClusters()
+        await Utils.delay(10000)
+      } else if (status === 'Creating') {
+        await Utils.delay(15000)
+      } else {
+        await Utils.delay(3000)
+      }
+    }
   }
 
-  componentDidUpdate() {
+  async componentDidMount() {
+    this.mounted = true
     this.loadIframe()
   }
 
   componentWillUnmount() {
+    this.mounted = false
     if (this.scheduledRefresh) {
       clearTimeout(this.scheduledRefresh)
     }
   }
 
   render() {
-    const { clusters } = this.props
-    const { url } = this.state
+    const { cluster } = this.props
+    const { url, createRequested } = this.state
+    const clusterStatus = cluster && cluster.status
 
-    return clusters && url ?
-      iframe({
+    if (clusterStatus === 'Running' && url) {
+      return iframe({
         src: url,
         style: {
-          border: 'none', flex: 1, marginBottom: '-1.5rem',
+          border: 'none', flex: 1,
           marginTop: -45, clipPath: 'inset(45px 0 0)' // cuts off the useless Jupyter top bar
         }
-      }) :
-      centeredSpinner()
+      })
+    } else {
+      return div({ style: { padding: '2rem' } }, [
+        spinner({ style: { color: colors.green[0], marginRight: '0.5rem' } }),
+        (clusterStatus === 'Creating' || !cluster || createRequested) ?
+          'Creating notebook runtime environment. You can navigate away and return in 5-10 minutes.' :
```

Comment:
Should these still say "notebook" for the terminal?

Reply:
```suggestion
          'Creating runtime environment. You can return in 5-10 minutes.' :
```
Same clarification as above